File: app/src/main/java/com/cmulugeta/mediaplayer/playback/QueueManager.java
package com.cmulugeta.mediaplayer.playback;


import android.support.annotation.NonNull;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaSessionCompat;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class QueueManager {

    private int currentIndex;
    private MetadataUpdateListener updateListener;
    private List<MediaSessionCompat.QueueItem> audioQueue;

    public QueueManager(@NonNull MetadataUpdateListener listener){
        this.updateListener=listener;
        this.audioQueue= Collections.synchronizedList(new ArrayList<MediaSessionCompat.QueueItem>());
        this.currentIndex=0;
    }

    public void setCurrentIndex(int index){
        if(checkForRange(index)){
            this.currentIndex=index;
        }
    }

    public boolean skipQueuePosition(int amount){
        final int index=currentIndex+amount;
        this.currentIndex=index<0||audioQueue==null?index:index % audioQueue.size();
        return checkForRange(index);
    }

    public MediaSessionCompat.QueueItem getCurrent(){
        return checkForRange(currentIndex)?audioQueue.get(currentIndex):null;
    }

    public int size(){
        return audioQueue!=null?audioQueue.size():0;
    }

    public void setCurrentItem(String mediaId){

    }

    public void updateMetadata(){
        MediaSessionCompat.QueueItem currentItem=getCurrent();
        if(currentItem!=null){
            updateListener.onMetadataRetrieveError();
            return;
        }


    }

    private boolean checkForRange(int index){
        return !(index<0||audioQueue==null||audioQueue.size()<=index);
    }


    private interface MetadataUpdateListener{
        void onMetadataChanged(MediaMetadataCompat metadata);
        void onMetadataRetrieveError();
        void onCurrentQueueIndexUpdated(int queueIndex);
        void onQueueUpdated(String title, List<MediaSessionCompat.QueueItem> newQueue);
    }
}

File: app/src/main/java/com/cmulugeta/mediaplayer/domain/playback/QueueManager.kt
package com.cmulugeta.mediaplayer.domain.playback

import com.cmulugeta.mediaplayer.domain.model.Track
import com.cmulugeta.mediaplayer.then
import java.util.Collections

class QueueManager(private var tracks: MutableList<Track>, var index: Int) {

  fun next() = !hasNext() then (current()) ?: tracks[++index]

  fun previous() = !hasPrevious() then (current()) ?: tracks[--index]

  fun shuffle() = Collections.shuffle(tracks)

  fun hasNext() = tracks.size > (index + 1)

  fun hasPrevious() = index - 1 >= 0

  fun size() = tracks.size

  fun addTrack(track: Track) = tracks.add(track)

  fun current() = tracks[index]
}